spec/unit/x509/cert_provider_spec.rb

```diff
@@ -0,0 +1,305 @@
+require 'spec_helper'
+require 'puppet/x509'
+
+describe Puppet::X509::CertProvider do
+  include PuppetSpec::Files
+
+  def pem(name)
+    File.read(my_fixture(name))
+  end
+
+  def cert(name)
+    OpenSSL::X509::Certificate.new(pem(name))
+  end
+
+  def crl(name)
+    OpenSSL::X509::CRL.new(pem(name))
+  end
+
+  def key(name)
+    OpenSSL::PKey::RSA.new(pem(name))
+  end
+
+  def request(name)
+    OpenSSL::X509::Request.new(pem(name))
+  end
+
+  def create_provider(options)
+    described_class.new(options)
+  end
+
+  context 'when loading' do
+    context 'cacerts' do
+      it 'returns nil if it does not exist' do
+        provider = create_provider(capath: '/does/not/exist')
+
+        expect(provider.load_cacerts).to be_nil
+      end
+
+      it 'returns an array of certificates' do
+        certs = create_provider(capath: my_fixture('ca.pem')).load_cacerts
+        expect(certs.count).to eq(1)
+        expect(certs.first.subject.to_s).to eq('/CN=Test CA')
+      end
+
+      it 'raises when invalid input is inside BEGIN-END block' do
+        ca_path = tmpfile('invalid_cacerts')
+        File.open(ca_path, 'w') do |f|
+          f.write '-----BEGIN CERTIFICATE-----'
+          f.write 'whoops'
+          f.write '-----END CERTIFICATE-----'
+        end
+
+        expect {
+          create_provider(capath: ca_path).load_cacerts
+        }.to raise_error(OpenSSL::X509::CertificateError)
+      end
+    end
+
+    context 'crls' do
+      it 'returns nil if it does not exist' do
+        provider = create_provider(crlpath: '/does/not/exist')
+        expect(provider.load_crls).to be_nil
+      end
+
+      it 'returns an array of CRLs' do
+        crls = create_provider(crlpath: my_fixture('crl.pem')).load_crls
+        expect(crls.count).to eq(1)
+        expect(crls.first.issuer.to_s).to eq('/CN=Test CA')
+      end
+
+      it 'raises when invalid input is inside BEGIN-END block' do
+        pending('jruby NPE bug') if Puppet::Util::Platform.jruby?
+
+        crl_path = tmpfile('invalid_crls')
+        File.open(crl_path, 'w') do |f|
+          f.write '-----BEGIN X509 CRL-----'
+          f.write 'whoops'
+          f.write '-----END X509 CRL-----'
+        end
+
+        expect {
+          create_provider(crlpath: crl_path).load_crls
+        }.to raise_error(OpenSSL::X509::CRLError, 'nested asn1 error')
+      end
+    end
+  end
+
+  context 'when saving' do
+    context 'cacerts' do
+      let(:ca_path) { tmpfile('pem_cacerts') }
+      let(:ca_cert) { cert('ca.pem') }
+
+      it 'writes PEM encoded certs' do
+        create_provider(capath: ca_path).save_cacerts([ca_cert])
+
+        expect(File.read(ca_path)).to match(/\A-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----\Z/m)
+      end
+
+      it 'sets mode to 644'
+    end
```

Comment:
Also, probably a good idea to define what happens if it fails to save a cacert

Reply:
And what happens if we have different content then an existing cert with the same name? Is it overwritten?

Reply:
I updated the CertProvider and tests to handle cases where we fail to save or load a cacert, private key, etc